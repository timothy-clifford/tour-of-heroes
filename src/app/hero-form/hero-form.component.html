<div class="container">
  <div *ngIf="!submitted">
    <form (ngSubmit)="onSubmit()" [formGroup]="heroForm">
      <div class="form-group">
        <label for="name">Name</label>
        <input type="text" class="form-control" id="name"
               required
               formControlName="name"
               name="name">
        <div *ngIf="name.invalid && name.dirty"
             class="alert alert-danger">
          Name is required
        </div>
      </div>

      <div class="form-group">
        <label for="alterEgo">Alter Ego</label>
        <input type="text" class="form-control" id="alterEgo"
               formControlName="alterEgo" name="alterEgo">
      </div>

      <div class="form-group">
        <label for="power">Hero Power</label>
        <select class="form-control" id="power"
                required
                formControlName="power"
                name="power">
          <option *ngFor="let pow of powers" [value]="pow">{{pow}}</option>
        </select>
        <div *ngIf="power.invalid && power.dirty" class="alert alert-danger">
          Power is required
        </div>
      </div>

      <div class="actions">
        <button type="submit" class="btn btn-success">Submit</button>
        <button type="reset" class="btn btn-default" (click)="newHero()">Reset</button>
      </div>
    </form>
  </div>

  <div *ngIf="submitted">
    <h2>You submitted the following:</h2>
    <div class="row">
      <div class="col-xs-3">Name</div>
      <div class="col-xs-9">{{ model.name }}</div>
    </div>
    <div class="row">
      <div class="col-xs-3">Alter Ego</div>
      <div class="col-xs-9">{{ model.alterEgo }}</div>
    </div>
    <div class="row">
      <div class="col-xs-3">Power</div>
      <div class="col-xs-9">{{ model.power }}</div>
    </div>
    <br>
    <button class="btn btn-primary" (click)="submitted=false">Edit</button>
  </div>
</div>