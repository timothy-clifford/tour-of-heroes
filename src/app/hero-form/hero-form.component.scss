.form-group {
  margin-bottom: 1rem;
}

.form-control {
  display: block;
  width: 100%;
  padding: .375rem .75rem;
  font-size: 1rem;
  line-height: 1.5;
  color: #495057;
  background-color: #fff;
  background-clip: padding-box;
  border: 1px solid #ced4da;
  border-radius: .25rem;
  transition: border-color .15s ease-in-out,box-shadow .15s ease-in-out;
}

label {
  display: inline-block;
  margin-bottom: .5rem;
}

.ng-valid[required], .ng-valid.required  {
  border-left: 5px solid #42A948; /* green */
}

form.ng-dirty .ng-invalid:not(form)  {
  border-left: 5px solid #a94442; /* red */
}

.container {
  width: 50vw;
}

.alert {
    position: relative;
    padding: .75rem 1.25rem;
    margin-bottom: 1rem;
    border: 1px solid transparent;
    border-radius: .25rem;
    width: 100%;

  &-danger {
    color: #721c24;
    background-color: #f8d7da;
    border-color: #f5c6cb;
  }
}

.actions {
  display: flex;
  justify-content: space-between;
}