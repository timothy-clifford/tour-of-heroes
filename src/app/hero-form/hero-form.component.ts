import { Component, Input, OnInit } from '@angular/core';
import { FormControl, FormGroup } from '@angular/forms';
import { Observable } from 'rxjs';

import { Hero } from '../hero';
import { HeroService } from '../hero.service';

@Component({
  selector: 'app-hero-form',
  templateUrl: './hero-form.component.html',
  styleUrls: ['./hero-form.component.scss']
})
export class HeroFormComponent implements OnInit {
  @Input() model?: Hero;

  heroForm?: FormGroup;

  powers = [
    'Mind Control',
    'Elemental Control',
    'Telekinesis',
    'Under Water Control',
    'Super Strength',
    'Super Speed',
    'Healing',
    'Shapeshifting',
    'Invisibility',
    'Flight'
  ];


  submitted = false;

  get name() {
    return this.heroForm.get('name');
  }

  get alterEgo() {
    return this.heroForm.get('alterEgo');
  }

  get power() {
    return this.heroForm.get('power');
  }

  get sideKick() {
    return this.heroForm.get('sideKick');
  }

  constructor(private heroService: HeroService) {
  }

  ngOnInit(): void {
    if (!this.model) {
      this.model = <Hero>{
        name: '',
        alterEgo: '',
        power: ''
      };
    }
    this.heroForm = new FormGroup({
      name: new FormControl(this.model.name),
      alterEgo: new FormControl(this.model.alterEgo),
      power: new FormControl({value: this.model.power, disabled: this.model.sideKick }),
      sideKick: new FormControl(this.model.sideKick)
    })
  }

  onSubmit() { 
    this.model = <Hero>this.heroForm.value;
    console.log(this.model);
    const action: Observable<any> = this.model.id ? this.heroService.updateHero(this.model) :
      this.heroService.addHero(this.model);
    action.subscribe();
  }

  newHero() {
    this.heroForm.reset();
  }

  handleSideKickChange($evt) {
    this.power.patchValue('Super Strength');
    $evt?.target?.checked ? this.power.disable() : this.power.enable()
  }
}