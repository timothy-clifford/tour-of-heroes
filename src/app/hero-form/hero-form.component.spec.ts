import { ComponentFixture, TestBed } from '@angular/core/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { HeroService } from '../hero.service';

import { HeroFormComponent } from './hero-form.component';

describe('HeroFormComponent', () => {
  let component: HeroFormComponent;
  let fixture: ComponentFixture<HeroFormComponent>;

  beforeEach(async () => {

    const heroServiceMock = {
      updateHero: jest.fn(),
      addHero: jest.fn(),
    };

    TestBed.configureTestingModule({
      declarations: [ HeroFormComponent ],
      imports: [ FormsModule, ReactiveFormsModule ],
      providers: [{ provide: HeroService, useValue: heroServiceMock }]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(HeroFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match a snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
