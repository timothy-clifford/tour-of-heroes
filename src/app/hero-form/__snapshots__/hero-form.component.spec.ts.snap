// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroFormComponent should match a snapshot 1`] = `
<app-hero-form
  heroForm={[Function FormGroup]}
  heroService={[Function Object]}
  model={[Function Object]}
  powers={[Function Array]}
  submitted="false"
>
  <div
    class="container"
  >
    <div>
      <form
        class="ng-untouched ng-pristine ng-invalid"
        novalidate=""
      >
        <div
          class="form-group"
        >
          <label
            for="name"
          >
            Name
          </label>
          <input
            class="form-control ng-untouched ng-pristine ng-invalid"
            formcontrolname="name"
            id="name"
            name="name"
            required=""
            type="text"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="alterEgo"
          >
            Side kick
          </label>
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            formcontrolname="sideKick"
            id="sideKick"
            name="sideKick"
            type="checkbox"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="alterEgo"
          >
            Alter Ego
          </label>
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            formcontrolname="alterEgo"
            id="alterEgo"
            name="alterEgo"
            type="text"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="power"
          >
            Hero Power
          </label>
          <select
            class="form-control ng-untouched ng-pristine ng-invalid"
            formcontrolname="power"
            id="power"
            name="power"
            required=""
          >
            <option
              value="Mind Control"
            >
              Mind Control
            </option>
            <option
              value="Elemental Control"
            >
              Elemental Control
            </option>
            <option
              value="Telekinesis"
            >
              Telekinesis
            </option>
            <option
              value="Under Water Control"
            >
              Under Water Control
            </option>
            <option
              value="Super Strength"
            >
              Super Strength
            </option>
            <option
              value="Super Speed"
            >
              Super Speed
            </option>
            <option
              value="Healing"
            >
              Healing
            </option>
            <option
              value="Shapeshifting"
            >
              Shapeshifting
            </option>
            <option
              value="Invisibility"
            >
              Invisibility
            </option>
            <option
              value="Flight"
            >
              Flight
            </option>
          </select>
        </div>
        <div
          class="actions"
        >
          <button
            class="btn btn-success"
            type="submit"
          >
            Submit
          </button>
          <button
            class="btn btn-default"
            type="reset"
          >
            Reset
          </button>
        </div>
      </form>
    </div>
  </div>
</app-hero-form>
`;
